# <p style="font-family: 'Amiri'; font-size: 2rem; color: #755139FF; text-align: center; margin: 0; text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.3); background-color: #F2EDD7FF; padding: 10px; border-radius: 20px; border: 5px solid #755139FF; width:95%">Nguyễn Đức Nhật - 21059221</p>

# <p style="font-family: 'Amiri'; font-size: 2rem; color: #755139FF; text-align: left; margin: 0; text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.3); background-color: #F2EDD7FF; padding: 10px; border-radius: 20px; border: 5px solid #755139FF; width:95%">Câu 1: Lý thuyết và phân tích mô hình</p>

# <p style="font-family: 'Amiri'; font-size: 1rem; color: #755139FF; text-align: left; margin: 0; text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.3); background-color: #F2EDD7FF; padding: 10px; border-radius: 20px; border: 3px solid #755139FF; width:95%">a) Ý nghĩa và cách tính các độ đo.</p>

- **Accuracy**: Tỷ lệ số dự đoán đúng trên tổng số mẫu. 
  $$Accuracy = \frac{TP + TN}{TP + FP + TN + FN}$$

- **Precision**: Trong số mẫu được dự đoán là Positive, có bao nhiêu mẫu thật sự là Positive. 
  $$Precision = \frac{TP}{TP + FP}$$

- **Recall**: Trong số các mẫu thật sự là Positive, có bao nhiêu mẫu được mô hình phát hiện. 
  $$Recall = \frac{TP}{TP + FN}$$

- **F1-score**: Trung bình điều hòa giữa Precision và Recall. 
  $$F1 = 2 * \frac{Precision * Recall}{Precision + Recall}$$


# <p style="font-family: 'Amiri'; font-size: 1rem; color: #755139FF; text-align: left; margin: 0; text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.3); background-color: #F2EDD7FF; padding: 10px; border-radius: 20px; border: 3px solid #755139FF; width:95%">b) Tính toán các độ đo với số liệu.</p>

In [2]:

TP = 80
FP = 20
TN = 90
FN = 10

accuracy = (TP + TN) / (TP + FP + TN + FN)
precision = TP / (TP + FP)
recall = TP / (TP + FN)
f1_score = 2 * (precision * recall) / (precision + recall)

print(f"Accuracy: {accuracy:.2f}")
print(f"Precision: {precision:.2f}")
print(f"Recall: {recall:.2f}")
print(f"F1 Score: {f1_score:.2f}")



Accuracy: 0.85
Precision: 0.80
Recall: 0.89
F1 Score: 0.84


> **Nhận xét:** Mô hình có Accuracy khá cao (85%), Precision và Recall đều ở mức tốt. F1-score cân bằng hai yếu tố nên đánh giá toàn diện mô hình.

# <p style="font-family: 'Amiri'; font-size: 2rem; color: #755139FF; text-align: left; margin: 0; text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.3); background-color: #F2EDD7FF; padding: 10px; border-radius: 20px; border: 5px solid #755139FF; width:95%">Câu 2: Dự đoán giá trị</p>


# <p style="font-family: 'Amiri'; font-size: 1rem; color: #755139FF; text-align: left; margin: 0; text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.3); background-color: #F2EDD7FF; padding: 10px; border-radius: 20px; border: 3px solid #755139FF; width:95%">a) Tính các độ đo MAE, RMSE, R².</p>

In [4]:

import numpy as np
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

y_true = [3.0, 4.5, 5.0, 6.0, 3.5]
y_pred = [2.5, 4.0, 5.2, 5.5, 3.0]

mae = mean_absolute_error(y_true, y_pred)
mse = mean_squared_error(y_true, y_pred)
rmse = np.sqrt(mse)
r2 = r2_score(y_true, y_pred)

print(f"Mean Absolute Error: {mae:.2f}")
print(f"Mean Squared Error: {mse:.2f}")
print(f"Root Mean Squared Error: {rmse:.2f}")
print(f"R-squared: {r2:.2f}")


Mean Absolute Error: 0.44
Mean Squared Error: 0.21
Root Mean Squared Error: 0.46
R-squared: 0.82



# <p style="font-family: 'Amiri'; font-size: 1rem; color: #755139FF; text-align: left; margin: 0; text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.3); background-color: #F2EDD7FF; padding: 10px; border-radius: 20px; border: 3px solid #755139FF; width:95%">b) Nhận xét.</p>
### Nhận xét mô hình hồi quy dựa trên các độ đo

#### 1. MAE = 0.44 triệu VND
- Trung bình mỗi dự đoán lệch **0.44 triệu VND** so với giá trị thực tế.
- Đây là sai số **tương đối nhỏ**, cho thấy mô hình có khả năng **ước lượng khá chính xác giá nhà**.

#### 2. RMSE = 0.46 triệu VND
- RMSE phạt nặng hơn những sai số lớn so với MAE.
- Vì RMSE xấp xỉ MAE, điều này cho thấy **không có sai số quá lớn nào** trong các dự đoán.
- Mô hình **ổn định và không bị ảnh hưởng bởi outliers**.

#### 3. R² (R-squared) = 0.82
- Hệ số xác định R² = 0.82 nghĩa là **82% phương sai của dữ liệu đầu ra (giá nhà)** được mô hình giải thích bằng các đặc trưng đầu vào (diện tích, số phòng).
- Đây là một giá trị **cao**, cho thấy mô hình **phù hợp khá tốt với dữ liệu**.

---

> ### Kết luận tổng thể:
- Mô hình có **hiệu suất dự đoán tốt**, sai số nhỏ, ổn định, và giải thích được phần lớn biến động giá nhà.
- Với **MAE & RMSE đều nhỏ hơn 0.5 triệu VND** và **R² > 0.8**, mô hình này **đủ tin cậy để sử dụng trong thực tế** *(đặc biệt nếu mức giá nhà trung bình cao hơn nhiều so với sai số này)*.



# <p style="font-family: 'Amiri'; font-size: 2rem; color: #755139FF; text-align: left; margin: 0; text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.3); background-color: #F2EDD7FF; padding: 10px; border-radius: 20px; border: 5px solid #755139FF; width:95%">Câu 3: Ứng dụng độ đo để chọn mô hình</p>

# <p style="font-family: 'Amiri'; font-size: 1rem; color: #755139FF; text-align: left; margin: 0; text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.3); background-color: #F2EDD7FF; padding: 10px; border-radius: 20px; border: 3px solid #755139FF; width:95%">a) Nếu ưu tiên phát hiện bệnh nhân càng nhiều càng tốt</p>

> Chọn mô hình M3 vì có **Recall = 0.95** cao nhất trong các mô hình.  

    - Điều này rất quan trọng trong bối cảnh y tế, vì giúp giảm tối đa khả năng bỏ sót bệnh nhân có bệnh 
    – tức là giảm số lượng False Negative, những người thực sự mắc bệnh nhưng bị mô hình dự đoán nhầm là không mắc.  

    - Một mô hình có Recall cao sẽ ưu tiên bắt được càng nhiều ca bệnh càng tốt, ngay cả khi phải đánh đổi bằng việc chấp nhận một vài cảnh báo sai (False Positive).


# <p style="font-family: 'Amiri'; font-size: 1rem; color: #755139FF; text-align: left; margin: 0; text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.3); background-color: #F2EDD7FF; padding: 10px; border-radius: 20px; border: 3px solid #755139FF; width:95%">b) Nếu muốn cân bằng giữa Precision và Recall</p>

> Chọn mô hình M3 vì có **F1-score = 0.92** cao nhất trong 3 mô hình.  

    - F1-score là độ đo tổng hợp giữa Precision và Recall, đặc biệt hữu ích khi cần cân bằng giữa việc phát hiện đúng bệnh nhân và tránh báo động giả.  

    - Giá trị F1-score cao cho thấy mô hình vừa chính xác khi phát hiện ca bệnh (Precision cao), vừa bắt được nhiều ca bệnh (Recall cao), giúp mô hình đáng tin cậy hơn trong thực tế y tế.


# <p style="font-family: 'Amiri'; font-size: 2rem; color: #755139FF; text-align: left; margin: 0; text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.3); background-color: #F2EDD7FF; padding: 10px; border-radius: 20px; border: 5px solid #755139FF; width:95%">Câu 4: Nhận diện lỗi và cải tiến mô hình</p>

# <p style="font-family: 'Amiri'; font-size: 1rem; color: #755139FF; text-align: left; margin: 0; text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.3); background-color: #F2EDD7FF; padding: 10px; border-radius: 20px; border: 3px solid #755139FF; width:95%">a) Vì sao Accuracy cao nhưng F1-score thấp?</p>

Mô hình có Accuracy rất cao nhưng F1-score lại thấp thường xảy ra khi **dữ liệu bị mất cân bằng nghiêm trọng**.

**Ví dụ**:  
- 95% mẫu là 'không bệnh', chỉ 5% là 'bệnh'.  
- Mô hình dự đoán toàn bộ là 'không bệnh' sẽ đạt **Accuracy = 95%**.  
- Nhưng lúc này, mô hình **không hề phát hiện bệnh nhân nào** → **Recall = 0**.  
- Precision và F1-score cũng thấp → cho thấy **mô hình không hữu ích cho lớp thiểu số**.

**F1-score thấp phản ánh rằng mô hình chưa làm tốt trong việc phân loại các trường hợp quan trọng (lớp thiểu số).**


# <p style="font-family: 'Amiri'; font-size: 1rem; color: #755139FF; text-align: left; margin: 0; text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.3); background-color: #F2EDD7FF; padding: 10px; border-radius: 20px; border: 3px solid #755139FF; width:95%">b) Các kỹ thuật cải thiện mô hình khi dữ liệu mất cân bằng</p>

>**1. Oversampling lớp thiểu số**  
- Tăng số lượng mẫu của lớp thiểu số để cân bằng tập dữ liệu.  
- Kỹ thuật phổ biến: **SMOTE (Synthetic Minority Over-sampling Technique)** — tạo mẫu mới bằng cách nội suy giữa các điểm gần nhau trong không gian đặc trưng.

>**2. Undersampling lớp chiếm đa số**  
- Giảm số lượng mẫu của lớp đa số để đạt cân bằng.  
- Có thể làm mất đi thông tin quan trọng nếu không cẩn thận.

>**3. Sử dụng `class_weight` khi huấn luyện mô hình**  
- Gán **trọng số lớn hơn cho lớp thiểu số** trong hàm mất mát.  
- Trong thư viện `scikit-learn`, có thể dùng `class_weight='balanced'` hoặc tự định nghĩa theo tỷ lệ mẫu.


# <p style="font-family: 'Amiri'; font-size: 1rem; color: #755139FF; text-align: left; margin: 0; text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.3); background-color: #F2EDD7FF; padding: 10px; border-radius: 20px; border: 3px solid #755139FF; width:95%">c) Vì sao Accuracy không phản ánh đúng chất lượng mô hình?</p>

Accuracy chỉ đo tỷ lệ tổng thể dự đoán đúng, nhưng **không quan tâm mô hình có làm tốt với lớp quan trọng hay không**.

**Ví dụ minh họa**:
- Có 1000 email, trong đó 950 là 'not spam', chỉ 50 là 'spam'.  
- Mô hình dự đoán tất cả là 'not spam'.  
  → **Accuracy = 950 / 1000 = 95%**  
  → **Recall = 0** đối với lớp spam (không phát hiện spam nào)  
  → **Precision undefined hoặc 0**  
  → **F1-score rất thấp**

**Kết luận**:  
Trong các bài toán **mất cân bằng dữ liệu**, độ chính xác cao **có thể gây hiểu nhầm**. Cần xem xét thêm các độ đo như **Precision, Recall và F1-score** để đánh giá mô hình một cách toàn diện.
